client/app/components/items-list/classes/ItemsSource.js

```diff
@@ -0,0 +1,247 @@
+import Paginator from './Paginator';
+import Sorter from './Sorter';
+import PromiseRejectionError from '@/lib/promise-rejection-error';
+
+class ItemsSource {
+  onBeforeUpdate = null;
+
+  onAfterUpdate = null;
+
+  onError = null;
+
+  getCallbackContext = () => null;
+
+  _beforeUpdate() {
+    if (isFunction(this.onBeforeUpdate)) {
+      return Promise.resolve(this.onBeforeUpdate(this.getState(), this.getCallbackContext()));
+    }
+    return Promise.resolve();
+  }
+
+  _afterUpdate() {
+    if (isFunction(this.onAfterUpdate)) {
+      return Promise.resolve(this.onAfterUpdate(this.getState(), this.getCallbackContext()));
+    }
+    return Promise.resolve();
+  }
+
+  // eslint-disable-next-line no-unused-vars
+  _changed(changes) {
+    this.update();
+  }
+
+  _getRequest() {
+    return this._originalGetRequest({
+      page: this._paginator.page,
+      page_size: this._paginator.itemsPerPage,
+      order: this._sorter.compiled,
+      q: isString(this._searchTerm) && (this._searchTerm !== '') ? this._searchTerm : undefined,
+      tags: this._selectedTags,
+    }, this.getCallbackContext());
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  _processResults(data) {
+    return data;
+  }
+
+  _updateResults(items, totalCount) {
+    this._pageItems = items;
+    this._paginator.setTotalCount(totalCount);
+  }
+
+  _doRequest(request) {
+    return this._beforeUpdate().then(() => (
+      this._originalDoRequest(request, this.getCallbackContext())
+        .then((data) => {
+          const { results, count } = this._processResults(data);
+          this._updateResults(results, count);
+          return this._afterUpdate();
+        })
+        .catch((error) => {
+          // ANGULAR_REMOVE_ME This code is related to Angular's HTTP services
+          if (error.status && error.statusText && error.data) {
+            error = new PromiseRejectionError(error.data.message);
+          }
+          this.handleError(error);
+        })
+    ));
+  }
+
+  constructor({ getRequest, doRequest, ...defaultState }) {
+    if (!isFunction(getRequest)) {
+      getRequest = identity;
+    }
+
+    this._originalGetRequest = getRequest;
+    this._originalDoRequest = doRequest;
+
+    this.setState(defaultState);
+    this._pageItems = [];
+  }
+
+  getState() {
+    return {
+      page: this._paginator.page,
+      itemsPerPage: this._paginator.itemsPerPage,
+      orderByField: this._sorter.field,
+      orderByReverse: this._sorter.reverse,
+      searchTerm: this._searchTerm,
+      selectedTags: this._selectedTags,
+      totalCount: this._paginator.totalCount,
+      pageItems: this._pageItems,
+    };
+  }
+
+  setState(state) {
+    this._paginator = new Paginator(state);
+    this._sorter = new Sorter(state);
+
+    this._searchTerm = state.searchTerm || '';
+    this._selectedTags = state.selectedTags || [];
+
+    this._savedOrderByField = this._sorter.field;
```

Comment:
I'm not sure, but wouldn't it be better to leave ItemResource out of it and have sorter manage (save/update/use) this property?

Reply:
It's the place where sorting logic was coupled with search. This field (and related functionality) are not directly related to sorting, that's some custom logic for a case when user searches something, and we should show results ordered by relevance. That's not something that Sorter should know about. Similar logic - when user selected tag - reset current page: Paginator shouldn't be coupled with tags, etc.

Reply:
And another possible question about this - why is this logic here and not in `ItemsList` component: because this class is framework-agnostic, it just encapsulates logic of fetching data and managing related state, but does not know how to render it; `ItemsList` just wraps both this "data source" together with target component (and adds some useful things, like adding route parameters) - we could do it in every list page we have, but this wrapper makes things easier.